### PR TITLE
RD-5479

### PR DESCRIPTION
**Pull request by OrrLumigo**
create extensions api in tracer

### CODE REVIEW

File: src/lumigo_tracer/lumigo_utils.py [hunk 1]
```diff
@@ -81,6 +82,9 @@
 internal_error_already_logged = False
 
 
+def get_use_tracer_extension() -> bool:
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:85]:
```suggestion
def should_use_tracer_extension() -> bool:
```

File: src/lumigo_tracer/lumigo_utils.py [hunk 2]
```diff
@@ -293,9 +298,12 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
     :param region: The region to use as default if not configured otherwise.
     :param msgs: the message to send.
     :param should_retry: False to disable the default retry on unsuccessful sending
+    :param use_extension: a flag to indicate if tracer should write spans to disk and let lumigo extension handle the send to lumigo
     :return: The duration of reporting (in milliseconds),
                 or 0 if we didn't send (due to configuration or fail).
     """
+    duration = 0
+    start_time = time.time()
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:306]: unnecessary changes

File: src/lumigo_tracer/lumigo_utils.py [hunk 3]
```diff
@@ -305,6 +313,13 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
     except Exception as e:
         get_logger().exception("Failed to create request: A span was lost.", exc_info=e)
         return 0
+    if get_use_tracer_extension():
+        file_name = f"{str(hashlib.md5(to_send))}_single"
+        file_path = f"/tmp/lumigo/lumigo-spans/{file_name}"
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:316]:
```suggestion
        file_path = os.path.join(EXTENSION_DIR, file_name)
```
export to const

File: src/lumigo_tracer/lumigo_utils.py [hunk 4]
```diff
@@ -305,6 +313,13 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
...
+        get_logger().info(f"writing spans to file {file_path}...")
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:317]:
```suggestion
        get_logger().info(f"writing spans to file {file_path}")
```

File: src/lumigo_tracer/lumigo_utils.py [hunk 5]
```diff
@@ -293,6 +298,7 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
     :param region: The region to use as default if not configured otherwise.
     :param msgs: the message to send.
     :param should_retry: False to disable the default retry on unsuccessful sending
+    :param use_extension: a flag to indicate if tracer should write spans to disk and let lumigo extension handle the send to lumigo
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:301]: ?

File: src/lumigo_tracer/lumigo_utils.py [hunk 6]
```diff
@@ -67,6 +67,7 @@
 NUMBER_OF_SPANS_IN_REPORT_OPTIMIZATION = 200
 DEFAULT_KEY_DEPTH = 4
 LUMIGO_TOKEN_KEY = "LUMIGO_TRACER_TOKEN"
+LUMIGO_USE_TRACER_EXTENSION = "LUMIGO_USE_TRACER_EXTENSION"
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:70]: you're not using it

File: src/lumigo_tracer/lumigo_utils.py [hunk 7]
```diff
@@ -305,6 +312,13 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
     except Exception as e:
         get_logger().exception("Failed to create request: A span was lost.", exc_info=e)
         return 0
+    if should_use_tracer_extension():
+        file_name = f"{str(hashlib.md5(to_send))}_single"
+        file_path = f"{EXTENSION_DIR}/{file_name}"
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:317]: pleaase use os.path.join

File: src/lumigo_tracer/lumigo_utils.py [hunk 8]
```diff
@@ -305,6 +312,14 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
     except Exception as e:
         get_logger().exception("Failed to create request: A span was lost.", exc_info=e)
         return 0
+    if should_use_tracer_extension():
+        file_name = f"{str(hashlib.md5(to_send).hexdigest())}_single"
+        Path(EXTENSION_DIR).mkdir(parents=True, exist_ok=True)
+        file_path = os.path.join(EXTENSION_DIR, file_name)
+        get_logger().info(f"writing spans to file {file_path}")
+        with open(file_path, "wb") as the_file:
+            the_file.write(to_send)
```
Comment (saartochner) [inline src/lumigo_tracer/lumigo_utils.py:321]: Maybe extract it to another function?

File: src/lumigo_tracer/lumigo_utils.py [hunk 9]
```diff
@@ -305,6 +312,9 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
     except Exception as e:
         get_logger().exception("Failed to create request: A span was lost.", exc_info=e)
         return 0
+    if should_use_tracer_extension():
+        write_spans_to_file(to_send)
+        return 0
```
Comment (nirLumigo) [inline src/lumigo_tracer/lumigo_utils.py:317]: This function suppose to return the duration of the reporting. Writing to file still takes some time, don't we want to measure it?

File: src/lumigo_tracer/lumigo_utils.py [hunk 10]
```diff
@@ -337,6 +347,15 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
     return duration
 
 
+def write_spans_to_file(to_send):
```
Comment (nirLumigo) [inline src/lumigo_tracer/lumigo_utils.py:350]:
```suggestion
def write_spans_to_file(to_send: bytes) -> None:
```

File: src/lumigo_tracer/lumigo_utils.py [hunk 11]
```diff
@@ -337,6 +347,15 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
...
+    file_name = f"{str(hashlib.md5(to_send).hexdigest())}_single"
```
Comment (nirLumigo) [inline src/lumigo_tracer/lumigo_utils.py:351]:
```suggestion
    file_name = f"{hashlib.md5(to_send).hexdigest()}_single"
```
Looks like `hexdigest` returns a string:
image

File: src/lumigo_tracer/lumigo_utils.py [hunk 12]
```diff
@@ -337,6 +347,15 @@ def report_json(region: Optional[str], msgs: List[dict], should_retry: bool = Tr
...
+    Path(EXTENSION_DIR).mkdir(parents=True, exist_ok=True)
+    file_path = os.path.join(EXTENSION_DIR, file_name)
+    get_logger().info(f"writing spans to file {file_path}")
+    with open(file_path, "wb") as spans_file:
+        spans_file.write(to_send)
```
Comment (nirLumigo) [inline src/lumigo_tracer/lumigo_utils.py:356]: I think that this part should be inside `lumigo_safe_execute` because accessing the files-system can result in an error.

File: src/test/unit/test_lumigo_utils.py [hunk 13]
```diff
@@ -442,6 +444,21 @@ def test_get_edge_host(arg, host, monkeypatch):
...
+    monkeypatch.setattr(Configuration, "should_report", True)
+    monkeypatch.setenv("LUMIGO_USE_TRACER_EXTENSION", "TRUE")
+    single = [{"a": "b"}]
+    to_send = _create_request_body(single, True).encode()
+    md5str = str(hashlib.md5(to_send).hexdigest())
+    file_name = f"{md5str}_single"
+    file_path = f"/tmp/lumigo-spans/{file_name}"
+    duration = report_json(None, [{"a": "b"}])
+    span_from_file = json.load(open(file_path, "r"))
+    assert path.exists(file_path)
```
Comment (nirLumigo) [inline src/test/unit/test_lumigo_utils.py:457]: No need to check if the file exists, if it does't exist the previous line will result in an exception.

File: src/test/unit/test_lumigo_utils.py [hunk 14]
```diff
@@ -442,6 +444,21 @@ def test_get_edge_host(arg, host, monkeypatch):
...
+    assert duration == 0
```
Comment (nirLumigo) [inline src/test/unit/test_lumigo_utils.py:458]: Are you sure we want to return `0`? (I also wrote about it inside the function)

File: src/test/unit/test_lumigo_utils.py [hunk 15]
```diff
@@ -442,6 +444,21 @@ def test_get_edge_host(arg, host, monkeypatch):
...
+    assert span_from_file == single
+
+
```
Comment (nirLumigo) [inline src/test/unit/test_lumigo_utils.py:461]: I think it's worth to add a tests where the file exist before running `report_json`

**Review by nirLumigo — CHANGES_REQUESTED** [review]
Nice! :)

File: src/test/unit/extension/conftest.py [hunk 16]
```diff
@@ -31,6 +32,24 @@ def mock_linux_files():
...
+                shutil.rmtree(file_path)
+        except Exception as e:
+            print('Failed to delete %s. Reason: %s' % (file_path, e))
+
+
+@pytest.fixture(autouse=True)
+def extension_clean():
+    clean_folder('/tmp/lumigo-spans')
+    yield
+    clean_folder('/tmp/lumigo-spans')
```
Comment (nirLumigo) [inline src/test/unit/extension/conftest.py:51]: You can simplify it:  🙂 
```suggestion
    if os.path.exists('/tmp/lumigo-spans'):
        shutil.rmtree('/tmp/lumigo-spans')
    yield
```

**Review by nirLumigo — CHANGES_REQUESTED** [review]

**Review by nirLumigo — APPROVED** [review]
